File: SistemaDeNotas.c
#include <stdio.h>
#include <string.h>
#include <ctype.h>

#define MAX_ALUNOS 100
#define NUM_NOTAS 3

void cadastrarAluno(char nomes[][MAX_ALUNOS], float notas[][NUM_NOTAS], int *quantidade) {
  printf("=====================\n");
  printf("Cadastro de Aluno\n");
  printf("=====================\n");
  printf("Digite o nome do aluno: ");
  scanf(" %[^\n]", nomes[*quantidade]);

while (1){
  printf("-------------------------\n");
  
  printf("Informe a 1ª nota do aluno: ");
  scanf("%f", &notas[*quantidade][0]);
  
  printf("Informe a 2ª nota do aluno: ");
  scanf("%f", &notas[*quantidade][1]);
  
  printf("Informe a 3ª nota do aluno: ");
  scanf("%f", &notas[*quantidade][2]);
  
  if (notas[*quantidade][0] >= 0 &&
      notas[*quantidade][0] <= 10 &&
      notas[*quantidade][1] >= 0 &&
      notas[*quantidade][1] <= 10 &&
      notas[*quantidade][2] >= 0 &&
      notas[*quantidade][2] <= 10) {
    
    break;

  } else {
    printf("Notas inválidas. Digite novamente.\n");
  }
}
(*quantidade)++;
  printf("Aluno cadastrado com sucesso!\n");
}

void listarAlunos(char nomes[][MAX_ALUNOS], float notas[][NUM_NOTAS], int quantidade) {
  if (quantidade == 0) {
    printf("Nenhum aluno cadastrado.\n");
    return;
  }

  printf("====================\n");
  printf("Alunos cadastrados:\n");
  printf("====================\n");
  for (int i = 0; i < quantidade; i++) {
    printf("%d. %s - Notas: %.2f, %.2f, %.2f\n", i + 1, nomes[i], notas[i][0], notas[i][1], notas[i][2]);
  }
}

float calcularMedia(float notas[][NUM_NOTAS], int quantidade) {
  if (quantidade == 0) {
    printf("Nenhuma nota registrada.\n");
    return 0;
  }

  float soma = 0;
  int quantidadeNotas = 0;
  for (int i = 0; i < quantidade; i++) {
    for (int j = 0; j < NUM_NOTAS; j++) {
      soma += notas[i][j];
      quantidadeNotas++;
    }
  }

  float media = soma / quantidadeNotas;
  printf("--------------------\n");
  printf("A média da turma é: %.2f\n", media);
  return media;
}

void listarAprovados(char nomes[][MAX_ALUNOS], float notas[][NUM_NOTAS], int quantidade) {
  if (quantidade == 0) {
    printf("Nenhum aluno cadastrado.\n");
    return;
  }

  printf("====================\n");
  printf("Alunos aprovados:\n");
  printf("====================\n");
  int quantidadeAprovados = 0;
  for (int i = 0; i < quantidade; i++){
    float soma = 0;
    for (int j = 0; j < NUM_NOTAS; j++) {
      soma += notas[i][j];
    }
    float media = soma / NUM_NOTAS;
    if (media >= 7) {
      printf("%d. %s - Média: %.2f\n", i + 1, nomes[i], media);
      quantidadeAprovados++;
    }
  }

  if (quantidadeAprovados == 0) {
    printf("Nenhum aluno aprovado.\n");
  }
}

int contarAprovadosRecursivo(float notas[][NUM_NOTAS], int quantidade, int indice) {
  if (indice == quantidade) {
    return 0;
  }
  float soma = 0;
  for (int j = 0; j < NUM_NOTAS; j++) {
    soma += notas[indice][j];
  }
  float mediaAluno = soma / NUM_NOTAS;
  if (mediaAluno >= 7){
    return 1 + contarAprovadosRecursivo(notas, quantidade, indice + 1);
  } else {
    return contarAprovadosRecursivo(notas, quantidade, indice + 1);
  }
}

int buscaSequencial(char nomes[][MAX_ALUNOS], float notas[][NUM_NOTAS], int quantidade) {
  char nomeBusca[MAX_ALUNOS];
  printf("--------------------\n");
  printf("Digite o nome do aluno que deseja buscar: ");
  scanf("%s", nomeBusca);
  printf("--------------------\n");
  for (int i = 0; i < quantidade; i++) {
    if (strcmp(nomes[i], nomeBusca) == 0) {
      printf("Aluno encontrado: %s - Notas: %.2f, %.2f, %.2f\n", nomes[i], notas[i][0], notas[i][1], notas[i][2]);
      return i;
    }
  }
  printf("Aluno não encontrado.\n");
  return -1;
}

float somarNotasRecursiva(float notas[][NUM_NOTAS], int quantidade, int indice) {
  if (indice == quantidade) {
    return 0;
    }
    float somaAluno = 0;
    for (int i = 0; i < NUM_NOTAS; i++) {
      somaAluno += notas[indice][i];
    }
    return somaAluno +
      somarNotasRecursiva(notas,quantidade,indice + 1);
}

void estatisticas(float notas[][NUM_NOTAS], int quantidade) {
  if (quantidade == 0) {
    printf("Nenhuma nota registrada.\n");
    return;
  }

  float soma = somarNotasRecursiva(notas, quantidade, 0);
  int quantidadeNotas = quantidade * NUM_NOTAS;
  float media = soma / quantidadeNotas;
  int aprovados = contarAprovadosRecursivo(notas, quantidade, 0);
  float percentual = ((float)aprovados / quantidade) * 100;
  
  printf("--------------------\n");
  printf("Estatísticas da turma:\n");
  printf("--------------------\n");
  printf("Soma das notas: %.2f\n", soma);
  printf("Média da turma: %.2f\n", media);
  printf("Percentual de alunos aprovados: %.2f%%\n", percentual);
  printf("Quantidade de alunos aprovados: %d\n", aprovados);
}

void mediaPorAluno(float notas[][NUM_NOTAS], char nomes[][100], int quantidade) {
  if (quantidade == 0) {
    printf("Nenhum aluno cadastrado.\n");
    return;
  }

  printf("====================\n");
  printf("Média por aluno:\n");
  printf("====================\n");
  for (int i = 0; i < quantidade; i++) {
    float soma = 0;
    for (int j = 0; j < NUM_NOTAS; j++) {
      soma += notas[i][j];
    }
    float media = soma / NUM_NOTAS;
    printf("%d. Média do aluno %s: %.2f\n", i + 1, nomes[i], media);
  }
}

void funcionalidadesExtras(char nomes[][100],float notas[][NUM_NOTAS],int quantidade) {
  char opcao2[10];
  while (1) {
    printf("====================\n");
    printf("Funcionalidades Extras\n");
    printf("====================\n");

    printf("S1 - Busca sequencial\n");
    printf("S2 - Estatisticas Complementares\n");
    printf("S3 - Contagem de Aprovados\n");
    printf("S4 - Media por aluno\n");
    printf("S - Voltar ao menu principal\n");

    printf("Escolha uma opcao: ");
    scanf("%s", opcao2);
  for (int i = 0; opcao2[i] != '\0'; i++) {
    opcao2[i] = tolower(opcao2[i]);
    }
  if (strcmp(opcao2, "s1") == 0) {
  buscaSequencial(nomes,notas,quantidade);
  } else 
  if (strcmp(opcao2, "s2") == 0) {
    estatisticas(notas,quantidade);
  } else 
  if (strcmp(opcao2, "s3") == 0) {
    int aprovados = contarAprovadosRecursivo(notas,quantidade,0);

  printf("--------------------\n");
  printf("Quantidade de aprovados: %d\n",aprovados);
  } else 
  if (strcmp(opcao2, "s4") == 0) {
    mediaPorAluno(notas,nomes,quantidade);
  } else 
  if (strcmp(opcao2, "s") == 0) {
    return;
  } else {
  printf("Opcao invalida.\n");
    }
  }
}

int main() {
  char nomes[MAX_ALUNOS][100];
  float notas[MAX_ALUNOS][NUM_NOTAS];
  int quantidade = 0;
  while (1) {
    printf("====================\n");
    printf("Sistema de Notas\n");
    printf("====================\n");

    printf("1 - Cadastrar aluno e notas\n");
    printf("2 - Exibir alunos e notas\n");
    printf("3 - Calcular media\n");
    printf("4 - Listar alunos aprovados\n");
    printf("5 - Acessar submenu funcionalidades extras\n");
    printf("0 - Sair\n");

    int opcao;
      printf("Escolha uma opcao: ");
      scanf("%d", &opcao);
      if (opcao == 1) {
        if (quantidade < MAX_ALUNOS) {
          cadastrarAluno(nomes,notas,&quantidade);
        } else {
          printf("Limite de alunos atingido.\n");
        }
        } else 
        if (opcao == 2) {
            listarAlunos(nomes,notas,quantidade);
        } else 
        if (opcao == 3) {
          calcularMedia(notas,quantidade);
        } else 
        if (opcao == 4) {
          listarAprovados(nomes,notas,quantidade);
        } else 
        if (opcao == 5) {
          funcionalidadesExtras(nomes,notas,quantidade);
        } else 
        if (opcao == 0) {
          printf("Saindo do sistema...\n");
          break;
        } else {
            printf("Opcao invalida. Tente novamente.\n");
        }
    }
    return 0;
}
    

File: SistemaDeNotas.py
#Criação das listas.
nomes = []
notas = []

#Função que cadastra alunos e suas respectivas notas.
def cadastrarAluno(nomes, notas):
  print('====================')
  print('Cadastro de Aluno')
  print('====================')
  nome = input("Informe o nome do aluno: ")
  while True:
    print('--------------------')
    n1 = float(input("Informe a 1º nota do aluno (0 a 10): "))
    n2 = float(input("Informe a 2º nota do aluno (0 a 10): "))
    n3 = float(input("Informe a 3º nota do aluno (0 a 10): "))
    #Esse if garante que a nota informada vai estar entre 0 e 10.
    if n1 >= 0 and n1 <= 10 and n2 >= 0 and n2 <= 10 and n3 >= 0 and n3 <= 10:
      break
    else:
      print("Nota inválida. Informe notas entre 0 e 10.")
  #A função append adiciona o nome do aluno e suas notas nas listas correspondentes.
  nomes.append(nome)
  notas.append((n1, n2, n3))
  print(f"Aluno cadastrado.")

#Função que lista os alunos cadastrados e suas respectivas notas.
def listarAlunos(nomes, notas):
  if len(nomes) == 0:
    print("Nenhum aluno cadastrado.")
    return
  print('====================')
  print("Alunos cadastrados:")
  print('====================')
  #Percorre a lista de nomes e notas, exibindo-os na tela.
  for i in range(len(nomes)):
    print(f"{i + 1}. {nomes[i]} - Nota: {notas[i]}")

#função que calcula a média da turma.
def calcularMedia(notas):
  if len(notas) == 0:
    print("Nenhuma nota registrada.")
    return 0
  soma = 0
  quantidadeNotas = 0
  for notasAluno in notas:
    for nota in notasAluno:
      soma += nota
      quantidadeNotas += 1    
#Armazena a soma das notas e a quantidade de notas para calcular a média.
  media = soma / quantidadeNotas
  print('--------------------')
  print(f"A média da turma é: {media:.2f}")
  return media

#Função que lista os alunos aprovados, ou seja, aqueles que possuem média maior ou igual a 7.
def listarAprovados(nomes, notas):
  if len(nomes) == 0:
    print("Nenhum aluno cadastrado.")
    return
  quantidade = 0
  print('====================')
  print("Alunos aprovados:")
  print('====================')
#Percorre a lista, utiliza a função sum() para somar todas as notas de cada aluno e divide pelo numero de notas registradas.
  for i in range(len(nomes)):
    media = sum(notas[i]) / len(notas[i])
    if media >= 7:
      print(f"{nomes[i]}")
      quantidade += 1
  if quantidade == 0:
    print("Nenhum aluno aprovado.")

#Função que conta a quantidade de alunos aprovados de forma recursiva.
def contarAprovadosRecursivo(notas, indice):
  if indice == len(notas):
      return 0
  mediaAluno = sum(notas[indice]) / len(notas[indice])
  if mediaAluno >= 7:
      return 1 + contarAprovadosRecursivo(notas, indice + 1)
  return contarAprovadosRecursivo(notas, indice + 1)

def buscaSequencial(nomes, notas):
  print('--------------------')
  nomeProcurado = input("Informe o nome do aluno que deseja buscar: ")
  print('--------------------')
  for i in range(len(nomes)):
    if nomes[i].lower() == nomeProcurado.lower():
      print('====================')
      print(f"Aluno encontrado: {nomes[i]} - Nota: {notas[i]}")
      return i
  print("Aluno não encontrado.")
  return -1

def somarNotasRecursiva(notas, indice):
  if indice == len(notas):
      return 0
  somaAluno = sum(notas[indice])
  return somaAluno + somarNotasRecursiva(notas, indice + 1)

def estatisticas(notas):
  if len(notas) == 0:
    print('Nenhuma nota cadastrada.')
    return
  soma = somarNotasRecursiva(notas, 0)
  aprovados = contarAprovadosRecursivo(notas, 0)
  quantidadeNotas = len(notas) * 3
  media = soma / quantidadeNotas
  percentual = (aprovados / len(notas)) * 100
  print('====================')
  print("Estatísticas Complementares:")
  print('====================')
  print(f"Soma das notas: {soma:.2f}")
  print(f"Média da turma: {media:.2f}")
  print(f"Quantidade de aprovados: {aprovados}")
  print(f"Percentual de aprovação: {percentual:.2f}%")


def mediaPorAluno(notas, nomes):
  medias = []
  for nota in notas:
    media = sum(nota) / len(nota)
    medias.append(media)
  print('--------------------')
  print("Médias dos alunos:")
  print('--------------------')
  for i in range(len(medias)):
    print(f"{i + 1}. {nomes[i]} - Média: {medias[i]:.2f}")

def funcionalidadesExtras(nomes, notas):
  while True:
    print('====================')
    print('Funcionalidades Extras')
    print('====================')
    print('S1 - Busca sequencial')
    print('S2 - Estatisticas Complementares')
    print('S3 - Contagem de Aprovados')
    print('S4 - Media por aluno')
    print('S - Voltar ao menu principal')
    opcao2 = input('Escolha uma opção: ').lower()
    if opcao2 == 's1':
      buscaSequencial(nomes, notas)
    elif opcao2 == 's2':
      estatisticas(notas)
    elif opcao2 == 's3':
      aprovados = contarAprovadosRecursivo(notas, 0)
      print('--------------------')
      print(f"Quantidade de aprovados: {aprovados}")
    elif opcao2 == 's4':
      mediaPorAluno(notas, nomes)
    elif opcao2 == 's':
      return
    else: 
      print("Opção inválida.")
      return

while True:
  print('====================')
  print('Sistema de Notas')
  print('====================')
  print('1 - Cadastrar aluno e notas')
  print('2 - Exibir alunos e notas')
  print('3 - Calcular média')
  print('4 - Listar alunos aprovados')
  print('5 - Acessar submenu funcionalidades extras')
  print('0 - Sair')
  opcao = input('Escolha uma opção: ')
  if opcao == '1':
    cadastrarAluno(nomes, notas)
  elif opcao == '2':
    listarAlunos(nomes, notas)
  elif opcao == '3':
    calcularMedia(notas)
  elif opcao == '4':
    listarAprovados(nomes, notas)
  elif opcao == '5':
    funcionalidadesExtras(nomes, notas)
  elif opcao == '0':
    print('Saindo do sistema...')
    break
  else:
    print('Opção inválida. Tente novamente.')
